plugins/plugin-ssh-machine/src/main/java/org/eclipse/che/plugin/machine/ssh/jsch/JschSshClient.java

```diff
@@ -0,0 +1,290 @@
+        }
+    }
+
+    private void copyFile(String sourcePath, String targetPath) throws MachineException {
+        ChannelSftp sftp = null;
+        try {
+            sftp = (ChannelSftp)session.openChannel("sftp");
+            sftp.connect(connectionTimeout);
+            String absoluteTargetPath = getAbsolutePath(targetPath);
+            copyFile(sourcePath, absoluteTargetPath, sftp);
+        } catch (JSchException e) {
+            throw new MachineException("Sftp copying failed. Error: " + e.getLocalizedMessage());
+        } finally {
+            if (sftp != null) {
+                sftp.disconnect();
+            }
+        }
+    }
+
+    private void copyFile(String sourcePath, String absoluteTargetPath, ChannelSftp channelSftp) throws MachineException {
+        try {
+            channelSftp.put(sourcePath, absoluteTargetPath);
+
+            // apply permissions
+            File file = new File(sourcePath);
+            int permissions = 4;
+            if (file.canExecute()) {
+                permissions += 1;
+            }
+            if (file.canWrite()) {
+                permissions += 2;
+            }
+            channelSftp.chmod(permissions, absoluteTargetPath);
+        } catch (SftpException e) {
+            throw new MachineException(format("Sftp copying of file %s failed. Error: %s",
+                                              absoluteTargetPath,
+                                              e.getLocalizedMessage()));
+        }
+    }
+
+    private String getAbsolutePath(String path) throws MachineException {
+        try {
+            return execAndGetOutput("cd " + path + "; pwd");
+        } catch (JSchException | IOException | MachineException e) {
+            throw new MachineException("Target directory lookup failed. " + e.getLocalizedMessage());
+        }
+    }
+
+    private int execAndGetCode(String command) throws JSchException {
+        ChannelExec exec = null;
+        try {
+            exec = (ChannelExec)session.openChannel("exec");
+            exec.setCommand(command);
+            exec.connect(connectionTimeout);
+
+            try {
+                // workaround exit code -1
+                Thread.sleep(1000);
+            } catch(InterruptedException ignored) {
+            }
+            return exec.getExitStatus();
+        } finally {
+            if (exec != null) {
+                exec.disconnect();
+            }
+        }
+    }
+
+    private String execAndGetOutput(String command) throws JSchException, MachineException, IOException {
+        ChannelExec exec = null;
+        try {
+            exec = (ChannelExec)session.openChannel("exec");
+            exec.setCommand(command);
+
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(exec.getInputStream()))) {
+
+                exec.connect(connectionTimeout);
+
+                try {
+                    // workaround exit code -1
+                    Thread.sleep(1000);
```

Comment:
I use commands such as "pwd" or "mkdir -p /some/path". And I need to wait to check if it return 0. If command is still running method returns -1.